// view-router/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const assert = require("assert");
const lodash_1 = require("lodash");
const path_1 = require("path");
const DEFAULT_CONFIG_PATH = "vrconfig.json";
class ViewRouter {
    constructor(views, options = {}) {
        this.options = options;
        this.views = new Map();
        for (const view of views) {
            assert.ok(view.id, "View has no id");
            assert.ok(view.urlPath, "View has no url path");
            if (view.urlPath.constructor === Array) {
                for (const path of view.urlPath) {
                    assert.equal(this.views.has(path), false);
                    this.views.set(path, view);
                }
            }
            else {
                assert.equal(this.views.has(view.urlPath), false);
                this.views.set(view.urlPath, view);
            }
        }
    }
    handle(req, res, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const view = this.views.get(req.path);
            if (!lodash_1.isNil(view)) {
                yield this.getView(req, res, view);
            }
            next();
        });
    }
    getView(req, res, viewConfig) {
        return __awaiter(this, void 0, void 0, function* () {
            let contextImport = null;
            try {
                contextImport = require(ViewRouter.getPath(lodash_1.isNil(viewConfig.viewHandlerPath) ? viewConfig.id : viewConfig.viewHandlerPath)).default;
            }
            catch (e) {
                console.warn(`Failed to import ${viewConfig.id}`);
            }
            const context = yield ((lodash_1.isNil(contextImport)) ?
                this.getDefaultView(req, res) :
                (contextImport.prototype.constructor === contextImport) ?
                    this.handleAsClass(contextImport, req, res) :
                    this.handleAsFunction(contextImport, req, res));
            return new Promise((resolve, reject) => {
                res.render((lodash_1.isNil(viewConfig.layout)) ? viewConfig.id : viewConfig.layout, context, (err, html) => {
                    if (err) {
                        return reject(err);
                    }
                    res.setHeader("Content-Type", "text/html; charset=utf-8");
                    res.send(html);
                    resolve();
                });
            });
        });
    }
    handleAsClass(viewConstructor, req, res) {
        return __awaiter(this, void 0, void 0, function* () {
            const view = new viewConstructor(req, res);
            if (!lodash_1.isNil(view.checkRouteValid) && (yield !view.checkRouteValid())) {
                // TODO add better clause, throw specific error
                // res.end();
                return {};
            }
            if (!lodash_1.isNil(this.options.basicContentGenerator) && !lodash_1.isNil(view.setBasicContext)) {
                view.setBasicContext(this.options.basicContentGenerator(req, res));
            }
            return view.getContext();
        });
    }
    handleAsFunction(handlerFunction, req, res) {
        return __awaiter(this, void 0, void 0, function* () {
            return handlerFunction(req, res);
        });
    }
    getDefaultView(req, res) {
        return {
            getContext: () => Promise.resolve((lodash_1.isNil(this.options.basicContentGenerator)) ? {} : this.options.basicContentGenerator(req, res)),
            setBasicContext: ctx => void 0,
            checkRouteValid: () => Promise.resolve(true)
        };
    }
    static getPath(subpath) {
        return path_1.join(this.path, subpath);
    }
}
ViewRouter.path = process.cwd();
function viewRouter(views, options) {
    if (!Array.isArray(views)) {
        options = views;
        views = undefined;
    }
    if (!lodash_1.isNil(options) && !lodash_1.isNil(options.basePath)) {
        ViewRouter.path = options.basePath;
    }
    if (lodash_1.isNil(views)) {
        let configReadPath = (!lodash_1.isNil(options) && !lodash_1.isNil(options.configFilePath)) ? options.configFilePath : DEFAULT_CONFIG_PATH;
        assert.equal(path_1.extname(configReadPath), ".json", "Not a json file");
        views = require(ViewRouter.getPath(configReadPath));
    }
    const vr = new ViewRouter(views, options);
    return (req, res, next) => {
        vr.handle(req, res, next);
    };
}
exports.viewRouter = viewRouter;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = viewRouter;
